src/utils.rs

```diff
@@ -35,6 +37,215 @@ use crate::FrameLayout;
 use crate::PlaneLayout;
 use crate::Resolution;
 
+/// A bit reader for h264 bitstreams. It properly handles emulation-prevention
+/// bytes and stop bits.
+pub(crate) struct NaluReader<'a> {
+    /// A reference into the next unread byte in the stream.
+    data: Cursor<&'a [u8]>,
+    /// Contents of the current byte. First unread bit starting at position 8 -
+    /// num_remaining_bits_in_curr_bytes.
+    curr_byte: u8,
+    /// Number of bits remaining in `curr_byte`
+    num_remaining_bits_in_curr_byte: usize,
+    /// Used in epb detection.
+    prev_two_bytes: u16,
+    /// Number of epbs (i.e. 0x000003) we found.
+    num_epb: usize,
+    /// Whether or not we need emulation prevention logic.
+    needs_epb: bool,
+}
+
+#[derive(Debug, Error)]
+pub(crate) enum GetByteError {
+    #[error("reader ran out of bits")]
+    OutOfBits,
+}
+
+#[derive(Debug, Error)]
+pub(crate) enum ReadBitsError {
+    #[error("more than 31 ({0}) bits were requested")]
+    TooManyBytesRequested(usize),
+    #[error("failed to advance the current byte")]
+    GetByte(#[from] GetByteError),
+    #[error("failed to convert read input to target type")]
+    ConversionFailed,
+}
+
+impl<'a> NaluReader<'a> {
+    pub fn new(data: &'a [u8], needs_epb: bool) -> Self {
+        Self {
+            data: Cursor::new(data),
+            curr_byte: Default::default(),
+            num_remaining_bits_in_curr_byte: Default::default(),
+            prev_two_bytes: 0xffff,
+            num_epb: Default::default(),
+            needs_epb: needs_epb,
+        }
+    }
+
+    /// Read a single bit from the stream.
+    pub fn read_bit(&mut self) -> Result<bool, ReadBitsError> {
+        let bit = self.read_bits::<u32>(1)?;
+        match bit {
+            1 => Ok(true),
+            0 => Ok(false),
+            _ => panic!("Unexpected value {}", bit),
+        }
+    }
+
+    /// Read up to 31 bits from the stream.
+    pub fn read_bits<U: TryFrom<u32>>(&mut self, num_bits: usize) -> Result<U, ReadBitsError> {
+        if num_bits > 31 {
```

Comment:
Just curious. Why not allow for up to 32? They would fit into `U`.

Reply:
It would break |read_se| since it would populate the signed bit of i32

Reply:
Added a comment describing this limitation.